### PR TITLE
DTS Motor Inverter Interface initial pull request

### PR DESCRIPTION
**Pull request by Burke-Daniel**
## DTS Motor Inverter Interface initial pull request
This PR contains the initial implementation of the DTS Interface, as well as the DTS motor inverter simulator, that we will use for testing. Other than a few small updates to the CanManager, and the can-sender and receiver examples, the main changes can be found in `Pidaq/Examples/dts_simulator/dts_simulator.py` and `Pidaq/Lib/dts_manager/dts_manager.py`.

### DTS Control
This class allows for control commands to be sent to the motor inverter, based on the CAN message format defined by the manufacturer

### DTS Telemetry
This class allows for the reception for (almost) all the information messages. The interface extracts the info from each can message byte, and converts it based off of the conversion factors outlined in the [CAN communications manual](https://sites.google.com/mun.ca/testing-software-wiki/documentation/Datasheets). All of the data fields are stored as member variables within the class. This code definitely needs a good refactor, but I'd like to get a bit of feedback first before that happens.

### DTS Simulator
This class utilizes the SocketCan linux protocol to send simulated Can messages that we would expect to receive from the real thing. The simulator will time out if it doesn't receive a control message from the interface every 0.5s. This code is not super clean, but functions as intended and is not production code anyways.

### Things left to do:
- Refactor DTS Manager after feedback
- docstrings and READMEs

### CODE REVIEW

File: Pidaq/Examples/dts_simulator/dts_simulator.py
Comment (colton-smith) [inline Pidaq/Examples/dts_simulator/dts_simulator.py]: Is there any specific reason that we don't use the CAN Manager class when creating this bus? I know it would be a lot of trouble to change this to get the same functionality.

File: Pidaq/Examples/dts_simulator/dts_simulator.py
Comment (colton-smith) [inline Pidaq/Examples/dts_simulator/dts_simulator.py]: Not saying change it but everywhere temperature was used for names, temp could probably have been used (temperature may  be a little verbose)

Comment (colton-smith) [reply Pidaq/Examples/dts_simulator/dts_simulator.py]: Looking back this might be used to reflect names in the datasheet - in that case I see why temperature could be better

File: Pidaq/Examples/dts_simulator/dts_simulator.py
Comment (colton-smith) [inline Pidaq/Examples/dts_simulator/dts_simulator.py:11]: I really like the simulator format, but I think it needs a way to display messages when they are received. 2 Ideas for this -either:
1. A  `BANGER` `__str__`  method that always us to call print(sim) and display some of the important info about the  current state of the simulated dts. This could just be things that we will commonly be sending like motor speed.

2. A method/methods called something along the lines of display_vitals() , display_temps(), etc. 

What I really want the implementation of this class to be capable of is:
-  Reading the canbus for a message (say an rpm command) 
-  Updating the rpm field based on the value specified in the message. 
-  Displaying a subset of the current status of the DTS (really only care about a few fields like speed)

File: Pidaq/Examples/dts_simulator/dts_simulator.py
Comment (colton-smith) [inline Pidaq/Examples/dts_simulator/dts_simulator.py]: Is it possible to add polling for speed commands via
```
if message.arbitration_id == 192:
            sim.read_configuration_message(message)
```

To this while loop? This way we can test sending multiple speed commands to the simulator. Also, it would be neat to add a call that prints the current info we care about (As mentioned in the initial comment about the class). 

I like the approach a lot though, we can configure a custom simulator that does only what we need using the 
provided methods. Very nice!

File: Pidaq/Lib/dts_manager/dts_manager.py
Comment (colton-smith) [inline Pidaq/Lib/dts_manager/dts_manager.py]: Not sure if it is just me but passing dicts around feels a little less secure/ deterministic than using an object or even just a ton of fields. 

If you want the values used in the function call to be portable (advantage of using a dict), it may be a good idea to have some sort of MotorConfig object and pass that instead.

File: Pidaq/Lib/dts_manager/dts_manager.py
Comment (colton-smith) [inline Pidaq/Lib/dts_manager/dts_manager.py]: Wondering if we could have a method that uses the configured values for everything except things related to speed / torque, but takes a DtsTestCommand object and uses the test_type field or whatever its called to set the mode for either an RPM or torque speed command. 

Basically, we have an initial configuration that sets everything to defaults and 0, then we send the same command with the required changes for speed / mode in this method.  

Also, we might need to consider having some sort of state flag mechanism. We could also consider placing the telemetry and control classes into a DTS class - I fear we may need the two to be very aware of each other.

For example, we will probably implement a method that checks the telemetry values and performs some action if the value exceeds a certain range. Trying to think about how we implement this - do we make it a utility function that takes a ref to the DtsControl / DtsTelemetry objects or would it be better suited to have a `master` class so to speak. 

Man. Design is hard.  Next working session we can talk about it

File: Pidaq/Lib/dts_manager/dts_manager.py
Comment (colton-smith) [inline Pidaq/Lib/dts_manager/dts_manager.py]: What is the purpose of this import?

File: Pidaq/Lib/dts_manager/dts_manager.py
Comment (colton-smith) [inline Pidaq/Lib/dts_manager/dts_manager.py]: Seeing as I think you have to change all the ids from hex strings to int now anyways, I would consider having a few fields for the important ids in the telemetry class instead of writing 0xa0/0xa1/0xa2/0xa3 that many times hahahah.

OR you could also add a convenience method like get_temp_data() or get_temp_msg() that returns self.bus.messages(WHATEVER ID). Up to you, either way it will be a lot easier to change an id if need be in the future

File: Pidaq/Lib/dts_manager/dts_manager.py
Comment (colton-smith) [inline Pidaq/Lib/dts_manager/dts_manager.py]: The slicing is very scary but looks correct. Might have been an idea to make a small convenience method to do this instead of repeating it - just a little less scary.

File: Pidaq/Lib/dts_manager/message_config.json
Comment (colton-smith) [inline Pidaq/Lib/dts_manager/message_config.json]: I guess this has to be updated to use integer message_ids now hey.... 
:(

File: Pidaq/Lib/dts_manager/dts_manager.py
Comment (colton-smith) [inline Pidaq/Lib/dts_manager/dts_manager.py]: What are digital inputs used for in the dts motor inverter?

File: Pidaq/Lib/dts_manager/dts_manager.py
Comment (colton-smith) [inline Pidaq/Lib/dts_manager/dts_manager.py]: Can this call be just bus.read_message_config('"dts", "message_config.json")

File: Pidaq/Lib/dts_manager/dts_manager.py
Comment (colton-smith) [inline Pidaq/Lib/dts_manager/dts_manager.py]: As I mentioned earlier - not sure if I like a dict or an object better for this. However, it seems like a lot of libraries do something similar to this (i.e. python-can) so maybe it isn't that bad...not sure hahaha

With an object at least you cant make typos. You could also implement restrictions on the config values if using a class.

File: Pidaq/Lib/dts_manager/dts_manager.py
Comment (colton-smith) [inline Pidaq/Lib/dts_manager/dts_manager.py]: Refer to previous comment about the sim not being configured to poll for motor commands, as well as the previous comment about the send_motor_command() not accepting new torque values.

**Review by colton-smith — CHANGES_REQUESTED** [review]
The structure/way that everything interfaces together is very promising! With a few changes, I think this is gonna be unreal. 

There are a few changes required that are related to:
- Now using integer and not hex values
- Need to be able to display values of the DTS Sim
- Need to be able to send motor speed messages without reconfiguring the entire motor config (if possible)

File: Pidaq/Examples/dts_simulator/dts_simulator.py
Comment (Burke-Daniel) [reply Pidaq/Examples/dts_simulator/dts_simulator.py]: Nope no reason not to be, and this is actually an easy change, as I can just call the underlying bus object in the `CanManager` class to send the messages, as here I want to be sending the already constructed `can.Messages` as opposed to using our send_message method and passing a list. Will make this change for sure

File: Pidaq/Examples/dts_simulator/dts_simulator.py
Comment (Burke-Daniel) [reply Pidaq/Examples/dts_simulator/dts_simulator.py]: Yeah, I was trying to keep the variable names as close to what they were listed on the datasheet

File: Pidaq/Examples/dts_simulator/dts_simulator.py
Comment (Burke-Daniel) [reply Pidaq/Examples/dts_simulator/dts_simulator.py]: Yup, meant to add that in there, that should be the behavior in that while loop so that the sim reacts to more speed commands.

File: Pidaq/Lib/dts_manager/dts_manager.py
Comment (Burke-Daniel) [reply Pidaq/Lib/dts_manager/dts_manager.py]: Good idea, similar to the SensorReading object in can_manager

File: Pidaq/Lib/dts_manager/dts_manager.py
Comment (Burke-Daniel) [reply Pidaq/Lib/dts_manager/dts_manager.py]: Great question... I have no idea hahaha, something to look into though

File: Pidaq/Lib/dts_manager/dts_manager.py
Comment (Burke-Daniel) [reply Pidaq/Lib/dts_manager/dts_manager.py]: indeed

File: Pidaq/Lib/dts_manager/dts_manager.py
Comment (Burke-Daniel) [reply Pidaq/Lib/dts_manager/dts_manager.py]: Yeah, went ahead and made a class and also some enums just to make the format of this a little stricter

File: Pidaq/Lib/dts_manager/dts_manager.py
Comment (Burke-Daniel) [reply Pidaq/Lib/dts_manager/dts_manager.py]: We didn't get a chance to talk about this on Sunday, but I figured I'd go ahead and throw something together for you to have a look at, I moved `DTSControl` and `DTSTelemetry` into a base `DTS` class where we can implement state and safety methods. Also went ahead and rejigged the motor command stuff, there are now Enums for a lot of the command stuff, and there are now defaults as well, I'll let you take a look and see what you think

File: Pidaq/Lib/dts_manager/dts_manager.py
Comment (Burke-Daniel) [reply Pidaq/Lib/dts_manager/dts_manager.py]: Lol that must've been there before I needed to import can for the actual class to make the can.Messages, removed

**Comment by Burke-Daniel** [comment]
@colton-smith I think I addressed all of your comments, definitely looks cleaner now! Back to you for another look

File: Pidaq/Examples/dts_simulator/dts_simulator.py
Comment (colton-smith) [inline Pidaq/Examples/dts_simulator/dts_simulator.py]: Swap the order of these 2 imports

File: Pidaq/Examples/dts_simulator/dts_simulator.py
Comment (colton-smith) [inline Pidaq/Examples/dts_simulator/dts_simulator.py]: Docstrings are like Russel Westbrook - Triple doubles """ """

File: Pidaq/Examples/dts_simulator/dts_simulator.py
Comment (colton-smith) [inline Pidaq/Examples/dts_simulator/dts_simulator.py:165]: Whats the reason behind dividing these guys by 10/100, just curious

File: Pidaq/Lib/dts_manager/dts_manager.py
Comment (colton-smith) [inline Pidaq/Lib/dts_manager/dts_manager.py]: Is inverter enable used as the heartbeat?

We never actually discussed if there is a dedicated CAN ID that serves as a heartbeat or if it can be any message from any ID. i.e. what do we need to send for the "heartbeat"

File: Pidaq/Lib/dts_manager/dts_manager.py
Comment (colton-smith) [inline Pidaq/Lib/dts_manager/dts_manager.py:387]: So this checks if there is any data, and the following calls parse specific bytes? Nice

File: Pidaq/Lib/dts_manager/message_config.json
Comment (colton-smith) [inline Pidaq/Lib/dts_manager/message_config.json:4]: Its cool how this file worked out so nicely

File: Pidaq/Lib/dts_manager/message_config.json
Comment (colton-smith) [inline Pidaq/Lib/dts_manager/message_config.json]: EOF newline

File: Pidaq/Lib/dts_manager/dts_manager.py
Comment (colton-smith) [inline Pidaq/Lib/dts_manager/dts_manager.py]: EOF newline? Weird github didnt warn about this one

File: Pidaq/Lib/dts_manager/dts_manager.py
Comment (colton-smith) [inline Pidaq/Lib/dts_manager/dts_manager.py]: Not sure if convert is the ideal name for these methods. They do convert the data from bytes to integers, but they also read the bus and update the class field. How do you feel about using update for these methods?

File: Pidaq/Lib/dts_manager/message_config.json
Comment (colton-smith) [inline Pidaq/Lib/dts_manager/message_config.json:7]: Am I correct in saying that for the inverter telemetry the conversion factor stored in the "Sensor Reading" object isn't actually used?

**Review by colton-smith — CHANGES_REQUESTED** [review]
Looks a lot cleaner than before. I like the use of enums, the motor config class, and the new way you parse the messages. In reality, we can only get so fancy! 

A few changes needed are:
- Triple double docstrings
- Couple EOF newlines
- Possibly use update instead of convert for the DTSTelem method names (up to you really, if that doesn't make sense to you then convert is fine)

The other things we talked about were:

- Moving the classes used for dts dash to the dts_manager package
This is something we can do after we merge this anyways - that can be a second tiny PR that gets merged instantly.

- Subclassing sensor reading object
Not sure if its necessary - we handle inverter conversions outside of the sensor reading class and sensor reading conversions in the class - if that gets too messy we can change it. 

Overall - there's not a whole lot at all that has to be changed before this can be merged imho - NICE JOB MATE

File: Pidaq/Examples/dts_simulator/dts_simulator.py
Comment (Burke-Daniel) [reply Pidaq/Examples/dts_simulator/dts_simulator.py:165]: So this is me just converting these back to the original values using the conversion factor, the values stored in the class already have the conversion factor applied (So for example a temp of 22.1 is stored as 221), just makes it easier to do it that way to pack them into bytes

File: Pidaq/Lib/dts_manager/dts_manager.py
Comment (Burke-Daniel) [reply Pidaq/Lib/dts_manager/dts_manager.py]: The "heartbeat" per se is actually the whole command message itself, so we are just going to have to continuously send out the command message every time we loop through the main

File: Pidaq/Lib/dts_manager/dts_manager.py
Comment (Burke-Daniel) [reply Pidaq/Lib/dts_manager/dts_manager.py:387]: Yeah I just have the default params for that to be 0 and 8, so that just checks to see if the whole message contains data, and if the message is `NoneType` then it will just skip over this. Worked out nicely

File: Pidaq/Lib/dts_manager/dts_manager.py
Comment (Burke-Daniel) [reply Pidaq/Lib/dts_manager/dts_manager.py]: Yeah, that probably describes them a bit better

File: Pidaq/Lib/dts_manager/message_config.json
Comment (Burke-Daniel) [reply Pidaq/Lib/dts_manager/message_config.json:7]: Yup, gonna change this so we are actually using them in the manager, and any that we don't need can be set to 1

**Comment by Burke-Daniel** [comment]
Alright, so I think I got everything, and added a few more nice features:
 - The periodic command message send like we discussed
 - Also created a method to update data based on what message id is received, so that we are not just calling each update method on every message received

File: Pidaq/Lib/can_manager/can_manager.py
Comment (colton-smith) [inline Pidaq/Lib/can_manager/can_manager.py]: Missing an if in this sentence lmao

**Review by colton-smith — APPROVED** [review]
Looks GTM - just a single typo. Look forward to implementing this stuff!

**Pull request merged**
